### DISCUSSION TITLE
Rerouting for island chains

### DISCUSSION THREAD

**Discussion by koriburkhardt** [jmlondon/pathroutr#20] [Q&A]
Hi there- I'm attempting to use your route_path() function for state space modeling of shark movement in French Polynesia. Is it possible to integrate shapefiles directly for the region as those in the package seem to not be detailed and accurate (the islands and atolls are very small and fine scale)

**Comment by jmlondon** [comment]
Thank you for the question.

`pathroutr` is capable of using any geographic source data you might have as long a you, first, convert to an `sf` object (e.g. with `sf::st_read()`). 

That said, each geographic data source is different and simply reading a shapefile in with `st_read()` might not _just work_ out of the box. The key issue is that the barrier data must be _a simple feature collection with geometry type 'POLYGON' or 'MULTIPOLYGON'._ There are some very basic checks in `pathroutr` that will look for these data types. But, they are pretty simple and don't provide much advice on how to fix.

My suggestion would be to start with something like the following:

```r
library(sf)
library(pathroutr)

# presuming you have a shapefile for the local region.
# note all of the other files associated w/ a shapefile should
# be present in the same directory as the *.shp file

my_barrier <- sf::st_read('path_to_my_shapefile.shp')
my_barrier <- sf::st_make_valid(my_barrier)

# presuming your track points/observed locations are also
# in a shapefile

my_pts <- sf::st_read('path_to_my_pts_shapefile.shp')

# my_pts and my_barrier must be in the same projection; use
# sf::st_transform() if needed

# create the visibility graph from your shapefile data

vis_graph <- prt_visgraph(my_barrier)

# re-route the tracks around the land barrier

track_pts_fix <- prt_reroute(my_pts, my_barrier, vis_graph)

# update the original data file with the re-routed locations

track_pts_fix <- prt_update_points(track_pts_fix, my_pts)

```

**Comment by koriburkhardt** [comment]
Thank you for getting back to me!
My track points are in a data frame rather than a shape file. I've
attempted to export the data frame as a shapefile (geometry is as points
rather than a connected polygon, so this is not preffered... still working
on this).

**Comment by koriburkhardt** [comment]
With both shapefiles formatted properly, running the vis_graph() portion
gives
warning messages:
1: In st_cast.sf(barrier, "POLYGON") :
   repeating attributes for all sub-geometries for which they may not be
constant
2: In st_triangulate.sfc(., bOnlyEdges = TRUE) :
   st_triangulate does not correctly triangulate longitude/latitude data

And if I continue to prt_reroute R aborts immediately.

**Comment by jmlondon** [comment]
A couple of things come to mind:

1. The warning you get regarding 'repeating attributes ...' shouldn't be of much concern. You can safely ignore this warning
2. The comment about 'st_triangulate does not correctly triangulate longitude/latitude data' is something you will need to address. Sorry for not emphasizing this in my previous response, but the input track data and the barrier polygon data both need to be in projected coordinates (not lat/long) and they both need to be in the same projection. The function `sf::st_transform()` will help you project your data but you'll need to know which projection is best for your specific region.

I'm hoping once you have the projections sorted, you'll be able to proceed without `prt_reroute()` aborting.

cheers
Josh

**Reply by kokoshark13** [reply]
Ah yes, that was the issue. Thank you for the clarification!

The problematic points have been rerouted, however they are displaced from the landmass points but are now crossing over adjacent barrier reefs and lagoons that are not possible for the species. Is there a way to include a buffer in this form such as in the `route_path` function?
![original paths](https://github.com/jmlondon/pathroutr/assets/122416726/d5bb5a4f-277c-4427-9809-94c4805c8590)
![Rerouted points](https://github.com/jmlondon/pathroutr/assets/122416726/4b3df6a3-3df3-457d-97dd-1c4b0c9f05fc)
